File: greedy/leetcode_680/solution.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        left_index = 0
        right_index = len(s) - 1
        
        def is_palindrome(string: str, left: int, right: int) -> bool:
            while left < right:
                if string[left] != string[right]:
                    return False
                
                left += 1
                right -= 1
            
            return True

        while left_index < right_index:
            if s[left_index] != s[right_index]:
                return is_palindrome(s, left_index + 1, right_index) or is_palindrome(s, left_index, right_index - 1)

            left_index += 1
            right_index -= 1

        return True

File: greedy/leetcode_680/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
var validPalindrome = function(s) {
    let leftIndex = 0;
    let rightIndex = s.length - 1;

    const isPalindrome = (str, left, right) => {
        while (left < right) {
            if (str[left] !== str[right]) {
                return false;
            }

            left ++;
            right --;
        }

        return true;
    }

    while (leftIndex < rightIndex) {
        if (s[leftIndex] !== s[rightIndex]) {
            return isPalindrome(s, leftIndex + 1, rightIndex) || isPalindrome(s, leftIndex, rightIndex - 1);
        }

        leftIndex ++;
        rightIndex --;
    }

    return true;
};